File: programmers/etc/level3/기지국_설치.py
from math import ceil

def solution(n, stations, w):
    answer = 0
    start = 1
    for station in stations:
        answer += max(ceil((station-w-start)/(2*w+1)), 0)
        start = station+w+1
    if n >= start:
        answer += ceil((n-start+1)/(2*w+1))
    return answer


n, station, w = 11, [4, 11], 1
print(solution(n, station, w)) # 3

n, station, w = 16, [9], 2
print(solution(n, station, w)) # 3

File: Java/programmers/etc/level3/기지국설치.java
package CodeTest.Java.programmers.etc.level3;

// Math.ceil은 시간초과 가능성이 있다.

public class 기지국설치 {
    public int solution(int n, int[] stations, int w) {
        int answer = 0;
        int start = 1;
        
        for (int station : stations) {
            int tmp = (station-w-start)/(2*w+1);
            if ((station-w-start)%(2*w+1)>0) tmp++; 
            answer += Math.max(tmp, 0);
            start = station+w+1;
        }
        
        if (n >= start) {
            int tmp = (n-start+1)/(2*w+1);
            if ((n-start+1)%(2*w+1)>0) tmp++;
            answer += tmp;
        }

        return answer;
    }
}
